### ISSUE TITLE
Study on cost benefit of running in GPU mode

### ISSUE THREAD

**Issue by SHuang-Broad** [google/deepvariant#650]
**Describe the issue:**

Hi, I am wondering if there's been a study on the cost-benefit of running DV in the GPU mode.
Back in the days of PEPPER-DeepVariant-Margin, I remember trying to profile (not as a rigorous study) what benefits there'd be if we were to run the pipeline in the GPU mode.
The conclusion back then from my anecdotal runs is that it's not worth it (we can get the CPU version to <$100/sample with little to minimum effort on optimizing cloud resource allocations, but the GPU version is ~$200/sample with P100).

Now given that DV has undergone quite a lot of changes since then, I wonder if the conclusion is changed.
So I did a test run on a PacBio Hifi 30X bam with DV 1.5.0, and collected the GPU resource log (using `gpustat -a -i 1 `, log attached below).
Looking at the log file, it doesn't look like GPU is used much still.

So I wonder if you have done any study on this subject and if so, can share some insights.

Thank you!

Steve

[gpu.usages.log.zip](https://github.com/google/deepvariant/files/11473421/gpu.usages.log.zip)

**Comment by AndrewCarroll** [comment]
Hi @SHuang-Broad 

This is a good question. There have been a number of changes to DeepVariant over releases. However, I don't think we've deeply revisited the timing advantages between CPU and GPU. I think that in our investigations, we see that in the GCP machines we use CPU instances still have some degree of cost advantage relative to GPU.

I agree that it's likely we're not optimally using the GPU, and I don't think we've specifically optimized for higher utilization of it. Although that's something we could look more into over time, right now we're prioritizing other feature improvements. 

For GPUs, how do you consider Parabricks as an option? I believe that with their most recent release, they will be able to run v1.5. For the optimization of GPU performance, I expect they are much better in using the GPU than our out-of-the-box code.

By the way, $100/sample seems high to me relative to our current experience. I hope that might be a function of the older method being Call-Phase-Call. If you have new benchmarks for this I would be curious to know.

Thanks,
Andrew

**Comment by pgrosu** [comment]
An interesting feature seems jump up when profiling of DV 1.5 in Docker, which looks to be heavy on I/O more than anything else:

![image](https://github.com/google/deepvariant/assets/6555937/a912b74c-3daa-44f1-a0d4-ef7ecc91bcce)

Then with sampling the I/O, the distribution of types of operations hint at some bottleneck opportunity:

![image](https://github.com/google/deepvariant/assets/6555937/9b6e2296-789b-4eff-852b-f053214f291a)

Maybe optimizing for I/O via balancing by having equal number of resources for `|GPU| == |CPU|` with significant on-board memory could do the trick, though not sure if all the code has been ported to Keras, which should help with the distribution strategy in that situation to make it all click together.

**Comment by SHuang-Broad** [comment]
Thank you both for your inputs.

Regarding the old, anecdotal observation that GPU isn't efficiently used, I didn't collect any GPU resources like this time.

And here are the CPU/memory/disk usages for the same input data, under CPU vs under GPU modes, back then.

Note these are collected from running the PEPPER pipeline (release 0.4.0). Note how using the GPU didn't help in terms wallclock time.
![pepper shard-0 high-cpu resources](https://github.com/google/deepvariant/assets/16310888/e7970a06-b38d-4972-8a8d-6a5cf0595ef1)
![pepper shard-0 NVDA-P100 resources](https://github.com/google/deepvariant/assets/16310888/78b9da04-d9c7-4c5b-a730-ae99b8ee4908)


And here's the cost of the GPU mode, broken down by SKU (we use Google cloud).

![pepper NVDA-P100 cost](https://github.com/google/deepvariant/assets/16310888/64316bdd-3ceb-47cb-a130-dd6da34dc8fc)

In terms of running DV in "production", we did optimize the CPU pipeline down to below $20/sample. We didn't optimize the GPU pipeline because of the wall-clock issue. Based on the samples I've analyzed with 1.5.0 so far, the cost indeed dropped.

Parabricks isn't really an option for us, as we build pipelines using WDL and run them on Terra.


With all these said, I'm definitely not saying that this is a blocker for us. It is more of a curiosity/optimization question.

Thank you!
Steve

**Comment by SHuang-Broad** [comment]
Thanks, Andrew!
I totally understand the priority setting.

In the meantime, I've collected some DV-1.5.0 resource usage data on a particular sample (note the scattering schemes aren't the same between the two references). Overall it's making really good use of the resources available (**_maybe_** except the memory, but it's more likely I over-allocated)

Hopefully, it can help your team with further tuning.
[30X.human.GRCh38.deepvariant.regular-resources-usage.pdf](https://github.com/google/deepvariant/files/11501324/30X.human.GRCh38.deepvariant.regular-resources-usage.pdf)
[30X.human.T2T.deepvariant.regular-resources-usage.pdf](https://github.com/google/deepvariant/files/11501325/30X.human.T2T.deepvariant.regular-resources-usage.pdf)

Thanks,
Steve

**Comment by pgrosu** [comment]
Hi Andrew and Steve,

That's a good idea, and probably shifting to pass-by-reference might minimize the overhead between functions or even through minimizing the number of functions.  A shared memory approach operated upon by the different functions could be ideal with maximizing performance, where even utilizing [NVidia's Unified Virtual Memory (UVM)](https://developer.nvidia.com/blog/maximizing-unified-memory-performance-cuda/) could help, but requires some care.  Currently the `call_variants` memory profile looks something like this, with minimal changes across the running process:

```
                                                           types |   # objects |   total size
================================================================ | =========== | ============
                                                             str |      223354 |     35.79 MB
                                                            dict |       88941 |     25.94 MB
                                                            code |       50107 |      8.54 MB
                                                            type |        6121 |      5.65 MB
                                                           tuple |       63884 |      3.62 MB
                                                            list |       30942 |      3.19 MB
                                                             set |        2864 |      1.51 MB
                                                         weakref |       14251 |   1002.02 KB
                                                     abc.ABCMeta |         784 |    826.05 KB
                                                            cell |       20911 |    816.84 KB
                                                             int |       25259 |    697.77 KB
                                      builtin_function_or_method |        8801 |    618.82 KB
  google.protobuf.pyext.cpp_message.GeneratedProtocolMessageType |         679 |    594.12 KB
                                                       frozenset |        1862 |    541.02 KB
                                             function (__init__) |        3439 |    456.74 KB
```

Based on the above, that made me curious as pass-by-reference could show benefits, but wanted to see what the functions with lots of calls might be.  Those types of functions are quite heavy regarding stack setup and teardown -- with intermediate local memory allocation -- which could also be the culprit based on Steve's temporal data:

```
Wed May 17 18:54:58 2023    call_variants_profile.txt

         12867185 function calls (11416794 primitive calls) in 61.905 seconds

   Ordered by: internal time
   List reduced from 3098 to 300 due to restriction <300>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        9   31.607    3.512   31.608    3.512 {built-in method tensorflow.python.client._pywrap_tf_session.TF_SessionRun_wrapper}
        2   14.137    7.069   14.137    7.069 {built-in method tensorflow.python.client._pywrap_tf_session.TF_CloseSession}
   434721    0.900    0.000    0.913    0.000 {built-in method builtins.hasattr}
   241294    0.868    0.000    3.111    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:5866(get_controller)
   241296    0.462    0.000    0.462    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/eager/context.py:932(_mode)
   413316    0.452    0.000    0.541    0.000 /usr/lib/python3.8/contextlib.py:82(__init__)
        9    0.409    0.045    0.409    0.045 {built-in method tensorflow.python.client._pywrap_tf_session.ExtendSession}
413315/164299    0.370    0.000    1.371    0.000 /usr/lib/python3.8/contextlib.py:117(__exit__)
825922/196    0.347    0.000   61.452    0.314 {built-in method builtins.next}
   113550    0.334    0.000    0.646    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:5887(_get_outer_context_and_inner_device_stack)
   111268    0.304    0.000    4.646    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:6030(executing_eagerly_outside_functions)
     1379    0.300    0.000    0.300    0.000 {method 'copy' of 'dict' objects}
   257584    0.293    0.000    0.293    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/eager/context.py:950(executing_eagerly)
413316/164111    0.253    0.000    2.725    0.000 /usr/lib/python3.8/contextlib.py:108(__enter__)
  659/110    0.247    0.000    0.248    0.002 /usr/local/lib/python3.8/dist-packages/tensorflow/python/keras/utils/version_utils.py:97(swap_class)
   413316    0.239    0.000    0.781    0.000 /usr/lib/python3.8/contextlib.py:238(helper)
     3367    0.228    0.000    0.228    0.000 {built-in method posix.stat}
     6015    0.223    0.000    0.223    0.000 {built-in method tensorflow.python.client._pywrap_tf_session.TF_FinishOperation}
   241296    0.210    0.000    0.240    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:5673(get_controller)
   256036    0.201    0.000    0.519    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/eager/context.py:2149(executing_eagerly)
    31845    0.189    0.000    0.189    0.000 {built-in method posix.lstat}
5154/4773    0.189    0.000    0.751    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/op_def_library.py:410(_ExtractInputsAndAttrs)
5154/4773    0.187    0.000    4.064    0.001 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/op_def_library.py:751(_apply_op_helper)
     6015    0.167    0.000    0.935    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:1906(_create_c_op)
627432/627423    0.165    0.000    0.182    0.000 {built-in method builtins.getattr}
709467/709457    0.165    0.000    0.232    0.000 {built-in method builtins.isinstance}
   564/94    0.149    0.000    0.149    0.002 /usr/local/lib/python3.8/dist-packages/keras/utils/version_utils.py:89(swap_class)
     6024    0.148    0.000    0.161    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/util/lock_util.py:105(_another_group_active)
   120647    0.140    0.000    0.266    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/eager/context.py:249(push)
   259701    0.137    0.000    0.197    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:5351(_device_function_stack)
    58954    0.117    0.000    0.200    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/dtypes.py:719(as_dtype)
     6669    0.113    0.000    0.118    0.000 /usr/local/lib/python3.8/dist-packages/gast/gast.py:15(create_node)
   155316    0.108    0.000    0.183    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:6314(get_default_graph)
     6015    0.105    0.000    0.105    0.000 {built-in method tensorflow.python.util._tf_stack.extract_stack_for_op}
     6015    0.104    0.000    0.149    0.000 /usr/local/lib/python3.8/dist-packages/tensorflow/python/framework/ops.py:1879(_NodeDef)
```

What I'm seeing above is something quite interesting.  The execution seems heavy on [eager execution rather than graph execution](https://blog.tensorflow.org/2018/08/code-with-eager-execution-run-with-graphs.html), which could be the culprit.  The thing is that there are some quick ways now to make it fast with almost no major code-rewrite, though there are other natural optimizations that are relatively obvious given the analysis.

For the `make_examples` analysis, the task is much simpler to fix by localizing I/O through just a shared memory model and/or pass-by-reference approach, as illustrated by the call distribution (with an eye on the serialization/deserialization of ProtoBuf if that becomes more heavily used):

```
Wed May 17 17:51:33 2023    make_examples_profile.txt

         2790682 function calls (2732833 primitive calls) in 11.860 seconds

   Ordered by: internal time
   List reduced from 11503 to 100 due to restriction <100>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  171/168    1.927    0.011    1.940    0.012 {built-in method _imp.create_dynamic}
     3159    0.930    0.000    0.930    0.000 {method 'read' of '_io.BufferedReader' objects}
    19018    0.694    0.000    0.694    0.000 {built-in method posix.stat}
        2    0.586    0.293    0.586    0.293 {method 'counts' of 'deepvariant.python.allelecounter.AlleleCounter' objects}
     3136    0.475    0.000    0.475    0.000 {built-in method marshal.loads}
     3136    0.422    0.000    0.422    0.000 {built-in method io.open_code}
    15152    0.183    0.000    0.183    0.000 {method 'reduce' of 'numpy.ufunc' objects}
     3136    0.164    0.000    2.265    0.001 <frozen importlib._bootstrap_external>:914(get_code)
     5151    0.146    0.000    0.436    0.000 /tmp/Bazel.runfiles_t7ewtxb0/runfiles/com_google_deepvariant/third_party/nucleus/util/genomics_math.py:196(normalize_log10_probs)
        1    0.146    0.146    0.146    0.146 {method 'summary_counts' of 'deepvariant.python.allelecounter.AlleleCounter' objects}
    10001    0.142    0.000    0.428    0.000 /tmp/Bazel.runfiles_t7ewtxb0/runfiles/com_google_deepvariant/deepvariant/variant_caller.py:249(with_gq_and_likelihoods)
     5151    0.134    0.000    0.610    0.000 /tmp/Bazel.runfiles_t7ewtxb0/runfiles/com_google_deepvariant/deepvariant/variant_caller.py:194(_calc_reference_confidence)
5268/5203    0.119    0.000    0.562    0.000 {built-in method builtins.__build_class__}
     1339    0.119    0.000    0.119    0.000 {method 'write' of 'third_party.nucleus.io.python.tfrecord_writer.TFRecordWriter' objects}
     1276    0.116    0.000    0.676    0.001 /tmp/Bazel.runfiles_t7ewtxb0/runfiles/com_google_deepvariant/deepvariant/variant_caller.py:221(make_gvcfs)
        2    0.107    0.054    0.107    0.054 {built-in method tensorflow.python.client._pywrap_tf_session.TF_LoadLibrary}
     5581    0.095    0.000    0.328    0.000 /usr/lib/python3.8/inspect.py:2124(_signature_from_function)
     4309    0.095    0.000    0.095    0.000 {method 'randint' of 'numpy.random.mtrand.RandomState' objects}
     5759    0.094    0.000    0.937    0.000 <frozen importlib._bootstrap_external>:1498(find_spec)
    23926    0.089    0.000    0.109    0.000 /usr/lib/python3.8/enum.py:313(__call__)
     3136    0.089    0.000    1.439    0.000 <frozen importlib._bootstrap_external>:1034(get_data)
      473    0.083    0.000    0.120    0.000 <frozen importlib._bootstrap_external>:1317(_path_hooks)
 1417/288    0.077    0.000    0.151    0.001 /usr/lib/python3.8/sre_parse.py:493(_parse)
     4910    0.069    0.000    0.069    0.000 {method 'PythonNext' of 'third_party.nucleus.io.python.sam_reader.SamIterable' objects}
   3614/1    0.069    0.000   11.965   11.965 {built-in method builtins.exec}
      701    0.066    0.000    0.066    0.000 {built-in method posix.listdir}
    15152    0.063    0.000    0.268    0.000 /usr/local/lib/python3.8/dist-packages/numpy/core/fromnumeric.py:69(_wrapreduction)
    32755    0.062    0.000    0.135    0.000 <frozen importlib._bootstrap_external>:121(_path_join)
  171/133    0.060    0.000    0.178    0.001 {built-in method _imp.exec_dynamic}
        1    0.059    0.059    0.059    0.059 {method 'calls_from_allele_counts' of 'deepvariant.python.variant_calling_multisample.VariantCaller' objects}
    23323    0.058    0.000    0.163    0.000 /usr/lib/python3.8/inspect.py:2489(__init__)
15563/15359    0.053    0.000    0.384    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
     3601    0.053    0.000    0.344    0.000 /usr/lib/python3.8/inspect.py:1102(getfullargspec)
    15152    0.051    0.000    0.319    0.000 /usr/local/lib/python3.8/dist-packages/numpy/core/fromnumeric.py:2703(amax)
    12155    0.049    0.000    0.060    0.000 {built-in method builtins.max}
    40200    0.049    0.000    0.049    0.000 {built-in method third_party.nucleus.util.python.utils.read_overlaps_region}
```

In any case, the most optimized approach is some sort shared memory model with minimal overhead to keep the code nimble and easy to integrate or extend.  The code itself is pleasantly easy to go through given the nice design, which makes it quite amendable to that.  In the meantime, pass-by-reference probably could be an easier win.  

Thank you,
Paul

**Comment by pichuan** [comment]
Thanks for all the feedback and discussion above! I'll close this issue now.
